### ISSUE TITLE
UX - Compare My Locations to Hotspots

### ISSUE THREAD

**Issue by iamjoshfrank** [revelry-foundation/nightingale#59]

**Comment by iamjoshfrank** [comment]
See [planning poker thread in Slack](https://revelry-public.slack.com/archives/C014AR56T1P/p1591038787332000) for deliberation.

Scoring a `5` (PokerBot chose not to work today).

Please keep in mind use of timeboxes for this effort. Thanks!

**Comment by HauwaAguillard** [comment]
picking this up next to pair with @DejaTrudeaux

**Comment by HauwaAguillard** [comment]
After pairing with @DejaTrudeaux and @Samkeer1 we found some code in the codebase that might have already solved for this. It looks like Joel's work from Friday, which brings down the complexity of work to be done on #44 

Here are some questions/concerns that came up during our pairing sessions

**What is considered a hotspot ?** 

- Covid19 +ive ? How many COVID 19 +ives ? is one enough to declare an area a hotspot?
- Could the busyness of a location trigger a hotspot?

**In terms of relevant data to the end-user, are we to determine hotspots based on:**
- locations visited i.e compare my locations visited hotspots
- locations closest to the end-user i.e As I move about my day, I expect hotspots to update based on proximity, at any given point in time.

### Potential Solutions
**Map**
Pending how we intend on showing hotspots i.e (locations visited or closest to me), we think it would take:
- One day lift to get a map
- One day lift to get points on the map
- half-day to style map pins

We are also aware we may get more hands-on-deck tomorrow and that could drastically reduce how long it takes to display a map.

**List**
Pending how we intend on showing hotspots i.e (locations visited or closest to me), we think it would be fairly straight forward to replicate work done on `locations`
for the hotspots list. 

**Filtering/Sorting**
We think sorting the data to what is most relevant to the end-user would be helpful either by time/location/distance

Next steps:
- Get clarifications on questions above ☝️
- Timebox the design efforts here for 1 hour or less

Reminders: 
Looks like most of the features here have been implemented by Joel. Below is a screenshot of a sample dataset for hotspots 

<img width="400" alt="Screen Shot 2020-06-01 at 3 39 30 PM" src="https://user-images.githubusercontent.com/49654454/83453805-bfae4780-a420-11ea-986b-6e659d8f8f4f.png">

@DejaTrudeaux @Samkeer1  feel free to add anything here I might have missed 🙏 

cc: @iamjoshfrank

**Comment by iamjoshfrank** [comment]
In general, please keep in mind the simplest solution for the current iteration.

> Covid19 +ive ? How many COVID 19 +ives ? is one enough to declare an area a hotspot?
> Could the busyness of a location trigger a hotspot?

For now, we should rely on whatever Joel put into the API to deem a match. I'm not concerned with, "is there 1 positive at the server or 5 positives for this given location match?" We can fine-tune this as we keep moving.

I feel like a lot of the questions posed in the comment are answered by, "What does the API / server give in response?" If you don't know / can't get an answer by inspecting the response or by looking at the relevant API code, then ping Joel to get an understanding.

**My BIG assumption:** the mobile portion for this feature is mostly a frontend:

Phone sends my local locations list to the server > Server feeds me back "matches" > phone displays matches to the user in whatever way makes sense _for the shape of the data right now._

**Comment by HauwaAguillard** [comment]
I see what you are saying and it does make sense. Not sure why I like to complicate things lol! 
I can certainly refocus my efforts on getting some visuals available on what is available. 

@Samkeer1 let me know if you're at a point where we could start putting existing data on a page to help guide the efforts here

**Comment by Samkeer1** [comment]
@iamjoshfrank @HauwaAguillard It looks like match is currently defined as Users within 10 meters and 30 minutes of where a Covid-19 Positive User has reported visiting. 

This seems to be the best way to determine a hotspot right off the bat: Where positive users have been. @jwietelmann built a function to check all positive locations against where a User has visited within the last 2 weeks.

**Comment by HauwaAguillard** [comment]
Thanks for clarifying @Samkeer1 sounds like we are comparing against a user's visited location. So I'm thinking this would be a lot similar to `locations` page and maybe with a button to run hotspots when a user requests for hotspots based on their visited locations. 

This unblocks me from being able to move forward with some wires. 

**My plan for tomorrow:** 
- Timebox this for 45 mins or less for some wires to help guide work done on #44. 
- I know @Samkeer1 plans to have some of the `Positive Match` data on the hotspot page and I believe design implementation can be done #44 

cc: @iamjoshfrank for visibility

**Comment by HauwaAguillard** [comment]
timeboxing this for 30mins to work in sketch.

**Comment by iamjoshfrank** [comment]
FYI from @jwietelmann [at end of All-Hands](fastlane increment_build_number): here's an example of a match between mobile and the server response:

https://nightingale-server.nightingale.revelry.org/api/v1/find_proximate_positives?lat=90&lng=90&when=2020-05-28T17:21:29Z

**Comment by HauwaAguillard** [comment]
Here are initial wires. Hoping to get some feedback

<img width="552" alt="Screen Shot 2020-06-02 at 10 50 12 AM" src="https://user-images.githubusercontent.com/49654454/83541582-719c5100-a4bf-11ea-8491-95fd60b08f51.png">

<img width="159" alt="Screen Shot 2020-06-02 at 10 54 54 AM" src="https://user-images.githubusercontent.com/49654454/83541635-82e55d80-a4bf-11ea-9e16-bf95239cba62.png">

cc: @Samkeer1 @iamjoshfrank

**Comment by Samkeer1** [comment]
@HauwaAguillard Looks good to me! Should there be a list specifically for comparing my precious locations to hotspots? That's how I currently have this working but not sure about how the wording should look, or where I should put this list. Maybe there could be a button on the hotspots page that takes someone to the list of where user locations overlapped with hotspots?

**Comment by iamjoshfrank** [comment]
Feedback incoming. Overall, this is a good direction for this first iteration of the feature.

Give me ~20 and I'll have my feedback documented.

**Comment by iamjoshfrank** [comment]
@HauwaAguillard - thank you! My feedback is below.

cc: @Samkeer1 since y'all are paired on the solution making.

---

**Panel 1**

I think this is solid. I'll provide content that might be needed, but this is a fine place to start.

I appreciate that we're not firing off the search query when the user navigates to this view. I like that it requires a discrete action and gives us an opportunity to reinforce the message that their own location history remains private and anonymous. Great job!

---

**Panel 2**

I think at this point we should consider dropping the explanatory text and replacing it with a subheading along the lines of:

> Possible Matches Based on Date, Time, and Location"

Alternatively, instead of a subheading, a content block with the same info in a more narrative format:

> Here is a list of possible matches between your locations and known hotspots. These possible matches are based on the date and time of visits to each location.

---

**Panels 2, 3, and 4**

I don't think we're creating any value by forcing the user to tap a location to see the details of the match. It's extra interaction to navigate back-and-forth between the list items and the detail view. And, the initial view on Panels 2,3 do not give the user a clear enough indication of:

- whether the locations in the list are their own locations or those of hotspots
- what to do next (tap to view the details)

I think a better solution might be a consolidation of the list such that Panels 2,3 look more like Panel 4.

It's an assumption, but I doubt in the real world that these matching lists are going to be extremely lengthy.

**Comment by iamjoshfrank** [comment]
> ... I was concerned about an extremely lengthy list ...
> ...
> I do agree that panel 4 definitely holds the most useful information, however, I think a workaround could be to show panel 4 as a top 5 location list and users could click on a view more link to view more?

Simple solution first. Show all matches. Pagination, details view, etc. are all solutions that can be explored when we verify the problem actually exists (of too many matches).

Also, I think we can greatly reduce the _size_ of these matches in order to make this view more user-friendly and scrollable. A lot can be done in terms of styling to make Panel 4 be the main list view including matches.

<img width="159" alt="Screen Shot 2020-06-02 at 10 54 54 AM" src="https://user-images.githubusercontent.com/49654454/83541635-82e55d80-a4bf-11ea-9e16-bf95239cba62.png">

**Comment by HauwaAguillard** [comment]
**Updated wires:** 
<img width="539" alt="Screen Shot 2020-06-02 at 12 58 08 PM" src="https://user-images.githubusercontent.com/49654454/83553374-b8467700-a4d0-11ea-94cd-7747815bbd7c.png">


cc: @Samkeer1 @iamjoshfrank

**Comment by iamjoshfrank** [comment]
Excellent!

**Comment by HauwaAguillard** [comment]
@iamjoshfrank I'm currently pairing with @Samkeer1 on design implementation on #44 and quick question. How often do we want to show the covid19 hotspots page or do we see that as a one-time show page for first time users?

**Comment by Samkeer1** [comment]
@HauwaAguillard @iamjoshfrank Can we just show matching locations in a list? To make hotspots readable in the same way that they're displayed in the wireframe, we need to send the following coord info through the google API, and there's nowhere currently in the DB to save those google API responses, meaning it would have to happen every time we display hotspots.

<img width="368" alt="Screen Shot 2020-06-02 at 2 42 23 PM" src="https://user-images.githubusercontent.com/43652523/83562492-46c1f500-a4df-11ea-9657-f91733ccb9fa.png">

**Comment by iamjoshfrank** [comment]
@Samkeer1 - start with what you are showing. For now, I don't care that the Hotspots are ugly - let's just prove we can retrieve them and display them alongside their matches.


My hunch about how we fix this (in a separate Issue):
- Update server to accept both the reverse geocoded info alongside lat/long when locations are sent to the server by positive users
- Update server's API response to hotspot matching queries to send back down the reverse geocoded positive locations

**Comment by iamjoshfrank** [comment]
@HauwaAguillard - I'm not sure I understand your question:

> How often do we want to show the covid19 hotspots page or do we see that as a one-time show page for first time users?

Can you restate it, maybe a bit differently?

**Comment by Samkeer1** [comment]
I think I understand, and @HauwaAguillard and Neil and I briefly discussed the fact that locations are already being passed through the google API locally for a user's local list of Locations so we can probably use those. PR should be incoming shortly

**Comment by HauwaAguillard** [comment]
@iamjoshfrank my question was in reference to this page, do we want this to be a first time user only or a page that persist each time a user visits the hotspot page
<img width="262" alt="Screen Shot 2020-06-02 at 3 06 14 PM" src="https://user-images.githubusercontent.com/49654454/83564617-9e159480-a4e2-11ea-85da-2b073aed8785.png">

**Comment by HauwaAguillard** [comment]
moving this to code review as I await feedback on the question above 👆

**Comment by iamjoshfrank** [comment]
@HauwaAguillard - now I understand!!

> do we want this to be a first time user only or a page that persist each time a user visits the hotspot page
> 
> <img width="262" alt="Screen Shot 2020-06-02 at 3 06 14 PM" src="https://user-images.githubusercontent.com/49654454/83564617-9e159480-a4e2-11ea-85da-2b073aed8785.png">

I think we should treat this view as a fresh search page and require the user to tap the "Compare my locations to Hotspots" button to submit a new search query any time the user returns to this view.

Here are a few reasons:

1. A user's locally-stored location history will be semi-constantly updating.
2. We might assume that a user will visit this Hotspot page multiple times, but probably on different days.
3. If a user visits the Hotspot page on Monday, and then again on Thursday, there will be new data to compare for matches. To keep things simple, and to reinforce the idea that we only use their local data for search queries (ie it's not being uploaded to a server in real-time and constantly analyzed), I think it helps to have this discrete CTA to tap to submit a Hotspot matching search.
4. If a user visits the Hotspot page multiple times in a single day, there will admittedly be some friction if we ask them to "tap the button to search for matches" each time. I think this is acceptable friction, especially for this Proof of Concept.

**Comment by HauwaAguillard** [comment]
ok sounds good. cc: @Samkeer1 for visibility.

**Comment by iamjoshfrank** [comment]
Bad assumptions re: what the server spits back as a result from the search request.

Joel is grabbing folks to clear it up.

**Comment by HauwaAguillard** [comment]
moving [this comment](https://github.com/revelry-foundation/nightingale/issues/44#issuecomment-637832586) over here for reference.

>Discussed the presentation of hotspot information with @HauwaAguillard, @jwietelmann and @NOrzechowski, and we agreed that the simplest way to tie off the core loop is to show text along the lines of 'x' amount of people tested positive, within the 'matching' constraints. @HauwaAguillard is working on mocking this up currently, while Neil and I work on displaying this data. This will allow us to iterate and further flesh out data that we're comfortable with displaying to user, as well as to iterate in terms of continuing to define what a hotspot is.Potentially, in further iterations it was mentioned that we could display more granular data to users. For example, 'the user was x distance away from you, for y amount of time'. etc.

**updated wires:** 

<img width="630" alt="Screen Shot 2020-06-02 at 5 20 15 PM" src="https://user-images.githubusercontent.com/49654454/83575680-b2af5800-a4f5-11ea-8a1b-c6541028ec91.png">

cc: @Samkeer1 @iamjoshfrank @jwietelmann @NOrzechowski for visibility

**Comment by iamjoshfrank** [comment]
👍  on this approach.

**Comment by HauwaAguillard** [comment]
Kewl. I'll update #44 with progress and move this ticket along the board
